File: backend/app/api/routes/resources.py
from typing import Any, cast

from fastapi import APIRouter, HTTPException, Query

from app.core.supabase import supabase
from app.services.ranking import compute_ranked_score

router = APIRouter()

SELECT = """
    id,
    created_at,
    title,
    description,
    resource_type,
    file_path,
    file_type,
    url,
    subjects ( id, name, subject_code, faculty_id, semester ),
    uploader:profiles ( id, full_name, reputation ),
    votes ( type, voter_id )
"""


@router.get(path="/ranked")
async def get_ranked_resources(
    faculty_id: str | None = Query(None),
    semester: int | None = Query(None),
    subject_id: str | None = Query(None),
    limit: int = Query(20, le=100),
    offset: int = Query(0, ge=0),
):
    query = supabase.table("resources").select(SELECT)

    if subject_id:
        query = query.eq("subject_id", subject_id)
    if faculty_id:
        query = query.eq("subjects.faculty_id", faculty_id)
    if semester is not None:
        query = query.eq("subjects.semester", semester)

    result = query.execute()
    resources = cast(list[dict[str, Any]], result.data or [])

    for r in resources:
        r["ranked_score"] = compute_ranked_score(r)
    resources.sort(key=lambda item: item["ranked_score"], reverse=True)

    return {
        "total": len(resources),
        "results": resources[offset : offset + limit],
    }


@router.get("/{resource_id}/score")
def get_resource_score(resource_id: str):
    result = (
        supabase.table("resources")
        .select(SELECT)
        .eq("id", resource_id)
        .maybe_single()
        .execute()
    )
    if not result:
        raise HTTPException(status_code=404, detail="Resource not found")

    resource = cast(dict[str, Any], result.data)

    return {"resource_id": resource_id, "ranked_score": compute_ranked_score(resource)}

File: frontend/src/api/resources.js
export const VOTE_POINTS = { UPVOTE: 1, DOWNVOTE: -1 };

export function computeScore(votes) {
  return votes.reduce((sum, v) => sum + (VOTE_POINTS[v.type] ?? 0), 0);
}

export async function fetchResourcesFeed(supabase) {
  const { data, error } = await supabase
    .from("resources")
    .select(`
      id,
      created_at,
      title,
      description,
      score,
      resource_type,
      file_path,
      file_type,
      url,
      subjects ( id, name, subject_code ),
      uploader:profiles ( id, full_name ),
      votes ( type, voter_id )
    `)
    .order("created_at", { ascending: false });

  if (error) throw error;
  return data;
}

export async function fetchSubjects(supabase, { facultyId, semester }) {
  const { data, error } = await supabase
    .from("subjects")
    .select("id, name")
    .eq("faculty_id", facultyId)
    .eq("semester", semester)
    .order("name");

  if (error) throw error;
  return data;
}

// Toggles a vote: same type again removes it, different type flips it,
// no existing vote inserts one. Returns the resulting vote type (or null)
// plus the score delta so the UI can update optimistically.
export async function castVote(supabase, { resourceId, voterId, type }) {
  const { data, error } = await supabase.rpc("toggle_vote", {
    p_resource_id: resourceId,
    p_voter_id: voterId,
    p_type: type,
  });
  if (error) throw error;
  return data;
}

export async function fetchResourceById(supabase, id) {
  const { data, error } = await supabase
    .from("resources")
    .select(`
      id,
      created_at,
      title,
      description,
      resource_type,
      file_path,
      file_type,
      url,
      subjects ( id, name, subject_code ),
      uploader:profiles ( id, full_name ),
      votes ( type, voter_id )
    `)
    .eq("id", id)
    .single();

  if (error) throw error;
  return data;
}

export async function fetchRelatedResources(supabase, { subjectId, excludeId }) {
  const { data, error } = await supabase
    .from("resources")
    .select(`
      id,
      title,
      votes ( type )
    `)
    .eq("subject_id", subjectId)
    .neq("id", excludeId)
    .limit(5);

  if (error) throw error;
  return data;
}
